### PR TITLE
Video: Add videos structured data

### LINKED ISSUE #12221: Add videos structured data
## Feature Description

Context: https://developers.google.com/search/docs/advanced/structured-data/video

This is basically like #11624, but for videos within a story.

We have a lot of information about videos already (URL, alt text, poster, duration, upload date) that is required or useful for `VideoObject`

I'd like to try this out for stories to see how well it works in WordPress

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Add video object to structured using for metadata. 

## Relevant Technical Choices

Feature is behind a feature flag. 

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Enable video meta experiment. 
2. Create a story. 
3. Insert a couple of videos. 
4. Insert a couple of products. 
5. Publish story. 
6. View story. 
7. View source.
8. Find `VideoObject`

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12221

### CODE REVIEW

File: includes/Model/Story.php [hunk 1]
```diff
@@ -190,8 +201,21 @@ public function load_from_post( $_post ): bool {
 		$this->markup  = $post->post_content;
 		$this->url     = (string) get_permalink( $post );
 
-		$thumbnail_id = (int) get_post_thumbnail_id( $post );
+		$this->load_poster_from_post();
+		$this->load_videos_from_post();
+		$this->load_publisher_logo_from_post();
+		$this->load_products_from_post();
```
Comment (spacedmonkey) [inline includes/Model/Story.php:207]: Broke up this method to keep PHPMD happy.

File: includes/Discovery.php [hunk 2]
```diff
@@ -272,13 +289,51 @@ protected function get_schemaorg_metadata(): array {
...
+		$video_data = [];
+		foreach ( $videos as $video ) {
+
+			$data = [
+				'@type'        => 'VideoObject',
+				'contentURL'   => $video->get_url(),
+				'name'         => $video->get_title(),
+				'description'  => $video->get_title(),
+				'thumbnailUrl' => $video->get_poster(),
+				'duration'     => sprintf( 'PT%dS', $video->get_duration() ),
```
Comment (spacedmonkey) [inline includes/Discovery.php:316]: Not sure that this is the best way to do this.

File: includes/Discovery.php [hunk 3]
```diff
@@ -272,13 +289,51 @@ protected function get_schemaorg_metadata(): array {
...
+			return [];
+		}
+		$video_data = [];
+		foreach ( $videos as $video ) {
+
+			$data = [
+				'@type'        => 'VideoObject',
+				'contentURL'   => $video->get_url(),
+				'name'         => $video->get_title(),
+				'description'  => $video->get_title(),
```
Comment (spacedmonkey) [inline includes/Discovery.php:314]: we need to work out what the description should say.

**Comment by spacedmonkey** [comment]
Example metadata. 

```json
{
  "@context": "http:\\/\\/schema.org",
  "publisher": {
    "@type": "Organization",
    "name": "Web Stories Dev"
  },
  "@type": "Article",
  "mainEntityOfPage": "http:\\/\\/localhost:8899\\/web-stories\\/832",
  "headline": "Test!!",
  "datePublished": "2022-09-29T16:58:29+08:00",
  "dateModified": "2022-09-29T16:59:24+08:00",
  "author": {
    "@type": "Person",
    "name": "admin"
  },
  "mainEntity": {
    "@type": "ItemList",
    "numberOfItems": "2",
    "itemListElement": [
      {
        "@type": "VideoObject",
        "contentURL": "http:\\/\\/localhost:8899\\/wp-content\\/uploads\\/2022\\/09\\/Lady-GaGa_Poker-Face-1-1_000.mp4",
        "name": "Lady-GaGa_Poker-Face-1-1_000",
        "description": "Lady-GaGa_Poker-Face-1-1_000",
        "thumbnailUrl": "http:\\/\\/localhost:8899\\/wp-content\\/uploads\\/2022\\/09\\/Lady-GaGa_Poker-Face-1-1_000-poster.jpeg",
        "duration": "PT36S",
        "uploadDate": "2022-09-28T15:46:03"
      },
      {
        "@type": "VideoObject",
        "contentURL": "http:\\/\\/localhost:8899\\/wp-content\\/uploads\\/2022\\/09\\/Lady-GaGa_Poker-Face-1-1_002.mp4",
        "name": "Lady-GaGa_Poker-Face-1-1_002",
        "description": "Lady-GaGa_Poker-Face-1-1_002",
        "thumbnailUrl": "http:\\/\\/localhost:8899\\/wp-content\\/uploads\\/2022\\/09\\/Lady-GaGa_Poker-Face-1-1_002-poster.jpeg",
        "duration": "PT37S",
        "uploadDate": "2022-09-28T15:44:55"
      }
    ]
  }
}
```

**Comment by swissspidy** [comment]
Hmm I don't think videos and products can be part of the same `ItemList` 🤔 

I need to look into how we could add both.

**Comment by spacedmonkey** [comment]
> Hmm I don't think videos and products can be part of the same `ItemList` 🤔
> 
> I need to look into how we could add both.

I run it through the validator. Seemed to work. 

Screenshot 2022-09-30 at 10 28 02

**Comment by swissspidy** [comment]
I will need to gather more information about this still.

**Comment by swissspidy** [comment]
I will look into it

**Comment by spacedmonkey** [comment]
@swissspidy Any word on this PR? It has been blocked for 2 weeks awaiting feedback.

**Comment by swissspidy** [comment]
No update yet, I will provide an update once I get to it and know more. It's not a critical PR, so no problem with it being open for a while.

**Pull request closed without merging**